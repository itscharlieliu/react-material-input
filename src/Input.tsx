import React, { ForwardedRef, MutableRefObject, useRef, useState } from "react";

import styles from "./styles/Input.module.css";

export interface InputProps extends React.InputHTMLAttributes<HTMLInputElement> {
    helperText?: string;
    label?: string;
    error?: boolean;
}

const Input = React.forwardRef(function Input(props: InputProps, ref: ForwardedRef<HTMLInputElement>): JSX.Element {
    const [value, setValue] = useState("");

    const { helperText, label, error, className, ...otherProps } = props;

    const inputRef = useRef<HTMLInputElement>(null);
    const hasValue = !!(inputRef.current && inputRef.current.value) || value;
    return (
        <label className={`${styles.input} ${error ? styles["input--error"] : ""} ${className ? className : ""}`}>
            <span className={`${styles.input__label} ${hasValue ? styles["input__label--minimized"] : ""}`}>
                {label}
            </span>
            <div className={`${styles.input__container}`}>
                <input
                    ref={(node: HTMLInputElement) => {
                        // Refs are not necessarily objects with a current property. They can also be functions.
                        // We need to write your code so that it can work with both variations.
                        (inputRef as MutableRefObject<HTMLDivElement | null>).current = node;
                        if (typeof ref === "function") {
                            ref(node);
                        } else if (ref) {
                            (ref as MutableRefObject<HTMLDivElement | null>).current = node;
                        }
                    }}
                    className={styles.input__element}
                    onChange={(event: React.ChangeEvent<HTMLInputElement>) => setValue(event.target.value)}
                    {...otherProps}
                />
            </div>
            <span className={styles.input__helperText}>{helperText}</span>
        </label>
    );
});

module.exports = Input;
