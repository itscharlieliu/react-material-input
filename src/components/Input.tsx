import React from "react";
import styles from "./styles/Input.module.css"

export interface InputProps extends React.InputHTMLAttributes<HTMLInputElement> {
    helperText?: string;
    label?: string;
    width?: number;
    error?: boolean;
}

// const Input = React.forwardRef(function Input(props: InputProps, ref: Ref<HTMLInputElement>): JSX.Element {
const Input = (): JSX.Element => {
    return (
        <label className={styles.inputContainer}>
            test
            <input/>
        </label>
    );
};

export default Input;