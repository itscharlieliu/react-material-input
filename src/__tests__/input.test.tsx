import { render, unmountComponentAtNode } from "react-dom";
import { act } from "react-dom/test-utils";
import renderer from "react-test-renderer";
import Input from "../Input";
import React from "react";

describe("input tests", () => {
    let container: null | HTMLDivElement = null;
    beforeEach(() => {
        // setup a DOM element as a render target
        container = document.createElement("input");
        document.body.appendChild(container);
    });

    afterEach(() => {
        // cleanup on exiting
        container && unmountComponentAtNode(container);
        container && container.remove();
        container = null;
    });

    it("renders without crashing", () => {
        act(() => {
            render(<Input />, container);
        });

        const component = renderer.create(<Input />).toJSON();
        expect(component).toMatchSnapshot("base");
    });

    it("renders when input contains text", () => {
        // TODO snapshot is not showing showing input label minimized
        const component = renderer.create(<Input value={"test"} />);
        expect(component).toMatchSnapshot("withValue");
    });
});
