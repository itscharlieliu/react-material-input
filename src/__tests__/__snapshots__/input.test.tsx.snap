// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`input tests renders when input contains text: withValue 1`] = `
<label
  className="input  "
>
  <span
    className="input__label "
  />
  <div
    className="input__container"
  >
    <input
      className="input__element"
      onChange={[Function]}
      value="test"
    />
  </div>
  <span
    className="input__helperText"
  />
</label>
`;

exports[`input tests renders without crashing: base 1`] = `
<label
  className="input  "
>
  <span
    className="input__label "
  />
  <div
    className="input__container"
  >
    <input
      className="input__element"
      onChange={[Function]}
    />
  </div>
  <span
    className="input__helperText"
  />
</label>
`;
